tests/e2e/tkgs_ha_utils.go

```diff
@@ -845,3 +804,116 @@ func exitHostMM(ctx context.Context, host *object.HostSystem, timeout int32) {
+	gomega.Expect(len(ssPodsBeforeScaleDown.Items) == int(replicas)).To(gomega.BeTrue(),
+		"Number of Pods in the statefulset should match with number of replicas")
+
+	ginkgo.By("Verify GV PV and SV PV has has required PV node affinity details")
+	ginkgo.By("Verify SV PVC has TKG HA annotations set")
+	// Get the list of Volumes attached to Pods before scale down
+	for _, sspod := range ssPodsBeforeScaleDown.Items {
+		pod, err := client.CoreV1().Pods(namespace).Get(ctx, sspod.Name, metav1.GetOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		for _, volumespec := range sspod.Spec.Volumes {
+			if volumespec.PersistentVolumeClaim != nil {
+				pvcName := volumespec.PersistentVolumeClaim.ClaimName
+				pv := getPvFromClaim(client, statefulset.Namespace, pvcName)
+				pvclaim, err := client.CoreV1().PersistentVolumeClaims(namespace).Get(ctx,
+					pvcName, metav1.GetOptions{})
+				gomega.Expect(pvclaim).NotTo(gomega.BeNil())
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				volHandle := getVolumeIDFromSupervisorCluster(pv.Spec.CSI.VolumeHandle)
+				gomega.Expect(volHandle).NotTo(gomega.BeEmpty())
+				svcPVCName := pv.Spec.CSI.VolumeHandle
+
+				svcPVC := getPVCFromSupervisorCluster(svcPVCName)
+				gomega.Expect(*svcPVC.Spec.StorageClassName == storagePolicyName).To(
+					gomega.BeTrue(), "SV Pvc storageclass does not match with SV storageclass")
+				framework.Logf("GC PVC's storageclass matches SVC PVC's storageclass")
+
+				verifyAnnotationsAndNodeAffinity(allowedTopologyHAMap, categories, pod,
+					nodeList, svcPVC, pv, svcPVCName)
+
+				// Verify the attached volume match the one in CNS cache
+				err = waitAndVerifyCnsVolumeMetadata4GCVol(volHandle, svcPVCName, pvclaim,
+					pv, pod)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+				framework.Logf(fmt.Sprintf("Verify volume: %s is attached to the node: %s",
+					pv.Spec.CSI.VolumeHandle, sspod.Spec.NodeName))
+				var vmUUID string
+				ctx, cancel := context.WithCancel(context.Background())
+				defer cancel()
+				vmUUID, err = getVMUUIDFromNodeName(pod.Spec.NodeName)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				verifyCRDInSupervisorWithWait(ctx, f, pod.Spec.NodeName+"-"+svcPVCName,
+					crdCNSNodeVMAttachment, crdVersion, crdGroup, true)
+
+				isDiskAttached, err := e2eVSphere.isVolumeAttachedToVM(client, volHandle, vmUUID)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				gomega.Expect(isDiskAttached).To(gomega.BeTrue(), "Disk is not attached to the node")
+				framework.Logf("verify the attached volumes match those in CNS Cache")
+				err = waitAndVerifyCnsVolumeMetadata4GCVol(volHandle, svcPVCName, pvclaim,
+					pv, pod)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			}
+		}
+	}
+
+}
+
+// Clean up statefulset and make sure no volume is left in CNS after it is deleted from k8s
+func cleaupStatefulset(client clientset.Interface, ctx context.Context, namespace string,
```

Comment:
This function can also be used for snapshot tests - if used, we need to add a logic to delete the snapshots for the volumes first and then delete volumes - can you add a //TODO section for code improvements needed in case if the function is called from snapshot test

Reply:
In that case this code can be moved to utils.go

Reply:
Moving this to utils.go and adding TODO for code improvement